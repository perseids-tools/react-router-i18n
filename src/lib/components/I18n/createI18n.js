import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import { matchPath, withRouter } from 'react-router-dom';

const hasKey = (obj, key) => (
  Object.prototype.hasOwnProperty.call(obj, key)
);

const dig = (obj, keys) => {
  if (!obj) {
    return null;
  }

  const [first, ...rest] = keys;

  if (rest.length === 0) {
    if (hasKey(obj, first)) {
      return obj[first];
    }

    return null;
  }

  if (hasKey(obj, first)) {
    return dig(obj[first], rest);
  }

  return null;
};

const getLocale = (path, defaultLocale, pathname) => {
  const { params } = matchPath(pathname, { path });

  if (!params) { return defaultLocale; }

  const { locale } = params;

  if (!locale) { return defaultLocale; }

  return locale;
};

const getText = (path, defaultLocale, pathname, t, children, translations, missingText) => {
  const locale = getLocale(path, defaultLocale, pathname);
  const splitT = t.split('.');

  if (hasKey(translations, locale)) {
    const translation = dig(translations[locale], splitT);

    if (translation !== null) {
      return translation;
    }
  }

  const defaultTranslation = dig(translations[defaultLocale], splitT);

  if (defaultTranslation !== null) {
    return defaultTranslation;
  }

  if (children !== undefined) { return children; }

  return missingText;
};

const createI18n = (locales, translations, missingText = false) => {
  const [defaultLocale] = locales;
  const path = `/:locale(${locales.join('|')})?`;

  const I18n = ({ location: { pathname }, children, t }) => (
    <Fragment>
      {getText(path, defaultLocale, pathname, t, children, translations, missingText)}
    </Fragment>
  );

  I18n.propTypes = {
    location: PropTypes.shape({
      pathname: PropTypes.string,
    }).isRequired,
    children: PropTypes.node,
    t: PropTypes.string.isRequired,
  };
  I18n.defaultProps = {
    children: undefined,
  };

  return withRouter(I18n);
};

export default createI18n;
