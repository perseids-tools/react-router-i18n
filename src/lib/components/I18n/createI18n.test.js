import React from 'react';
import renderer from 'react-test-renderer';
import { MemoryRouter } from 'react-router-dom';
import createI18n from './createI18n';

const locales = ['en', 'fr'];
let translations = {
  en: {
    home: 'Home',
  },
  fr: {
    home: 'Accueil',
  },
};

let I18n;
beforeEach(() => {
  I18n = createI18n(
    locales,
    translations,
  );
});

it('renders without crashing', () => {
  const component = (
    <MemoryRouter>
      <I18n t="home" />
    </MemoryRouter>
  );

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

it('shows the translation of another locale', () => {
  const component = (
    <MemoryRouter initialEntries={['/fr/']}>
      <I18n t="home" />
    </MemoryRouter>
  );

  const tree = renderer.create(component).toJSON();
  expect(tree).toMatchSnapshot();
});

describe('nested translations', () => {
  beforeAll(() => {
    translations = {
      en: {
        nested: {
          translation: 'Nested',
        },
      },
    };
  });

  it('shows a nested translation', () => {
    const component = (
      <MemoryRouter>
        <I18n t="nested.translation" />
      </MemoryRouter>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('function translations', () => {
  beforeAll(() => {
    translations = {
      en: {
        time: ({ hour, minute }) => `${hour}:${minute}`,
      },
    };
  });

  it('executes the function and uses the result as the translation', () => {
    const component = (
      <MemoryRouter>
        <I18n t="time" args={{ hour: '12', minute: '30' }} />
      </MemoryRouter>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});

describe('default translations', () => {
  beforeAll(() => {
    translations = {
      en: {
        home: 'Home',
      },
    };
  });

  it('uses the default language translation when there is none', () => {
    const component = (
      <MemoryRouter initialEntries={['/fr/']}>
        <I18n t="home" />
      </MemoryRouter>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('uses props.children when there is no translation', () => {
    const component = (
      <MemoryRouter>
        <I18n t="other">
          Other
        </I18n>
      </MemoryRouter>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });


  it('does not render when there is no translation', () => {
    const component = (
      <MemoryRouter>
        <I18n t="other" />
      </MemoryRouter>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('uses the default if one is given and there is no translation', () => {
    I18n = createI18n(
      locales,
      translations,
      'Missing',
    );

    const component = (
      <MemoryRouter>
        <I18n t="other" />
      </MemoryRouter>
    );

    const tree = renderer.create(component).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
