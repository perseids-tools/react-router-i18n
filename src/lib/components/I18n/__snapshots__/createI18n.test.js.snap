// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default translations does not render when there is no translation 1`] = `null`;

exports[`default translations uses props.children when there is no translation 1`] = `"Other"`;

exports[`default translations uses the default if one is given and there is no translation 1`] = `"Missing"`;

exports[`default translations uses the default language translation when there is none 1`] = `"Home"`;

exports[`nested translations shows a nested translation 1`] = `"Nested"`;

exports[`renders without crashing 1`] = `"Home"`;

exports[`shows the translation of another locale 1`] = `"Accueil"`;
